source/extensions/resource_monitors/unused_heap/unused_heap_monitor.cc

```diff
@@ -0,0 +1,50 @@
+#include "extensions/resource_monitors/unused_heap/unused_heap_monitor.h"
+
+#include "envoy/config/resource_monitor/unused_heap/v2alpha/unused_heap.pb.h"
+
+#include "common/common/assert.h"
+#include "common/memory/stats.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace UnusedHeapMonitor {
+
+UnusedHeapMonitor::UnusedHeapMonitor(
+    const envoy::config::resource_monitor::unused_heap::v2alpha::UnusedHeapConfig& config,
+    std::unique_ptr<Common::MemoryStatsReader> stats)
+    : config_(config), stats_(std::move(stats)) {}
+
+void UnusedHeapMonitor::updateResourceUsage(Server::ResourceMonitor::Callbacks& callbacks) {
+  const size_t physical = stats_->reservedHeapBytes();
+  const size_t unmapped = stats_->unmappedHeapBytes();
+  ASSERT(physical >= unmapped);
+  const size_t reserved = physical - unmapped;
+  const size_t allocated = stats_->allocatedHeapBytes();
+  ASSERT(reserved >= allocated);
+
+  Server::ResourceUsage usage;
+
+  switch (config_.monitor_type_case()) {
+  case envoy::config::resource_monitor::unused_heap::v2alpha::UnusedHeapConfig::MonitorTypeCase::
+      kMaxUnusedHeapSizeBytes: {
+    usage.resource_pressure_ =
+        (reserved - allocated) / static_cast<double>(config_.max_unused_heap_size_bytes());
```

Comment:
I'm not fully grokking the utility of this metric for overload manager. What is the intuition behind putting limits on the unused heap bytes?

Reply:
some discussions about this PR can be seen in the #9381 

I will find that Envoy uses a lot of memory in the actual production environment, but the actual allocated memory is not much, most of the memory is reserved by TCmalloc, so I need a mechanism to release this part of memory. The purpose of this PR is to monitor unused memory for memory reclamation

Reply:
I see. Is there no way to periodically shrink the heap with tcmalloc? What is best practice in other systems that work around this?

Reply:
Frankly, I have no experience in this respect, for other server-side programs, TCmalloc's reserved memory can partly improve the efficiency of memory allocation, if Envoy as a gateway's role in running, I'm willing to sacrifice some memory for performance, but as the sidecars, I don't want to take up too much memory, is can sacrifice some performance. Envoy already has a mechanism for memory shrink (fixed_heap), but it is not flexible enough to configure only a fixed memory size, which triggers the shrink when the memory reaches this value, but this value is hard to determine in production because the memory usage of each sidecar is different.

Reply:
Stepping back for a second, there is an underlying issue here in that Envoy is bursting to use "a lot" of memory in your production environment. IMO it would be better to constrain Envoy's memory usage with the various controls we have to do that vs. try to give the memory back when Envoy uses too much during these bursts, as there are likely to be times in which the burst will actually cause an OOM.

Also, I do wonder if it would be worth experimenting with a different allocator like jemalloc which anecdotally I think may more aggressively work to return memory to the OS than tcmalloc does.

Reply:
@mattklein123 
So far, it is not the Envoy's problem. There are two main scenarios that can cause the Envoy to retain more memory than it actually allocates.

1. We accidentally called the admin's '/cluster' interface in the production environment, which will dump all the cluster and IP information (cluster and IP are large in size in our production environment), which will result in a large amount of memory allocation.

2. Some bugs in our control plane caused a large cluster to be pushed by mistake, which caused the Envoy to allocate a large amount of memory. It would take some time to remove the cluster.

Reply:
Yeah, the other situation this can arise is if we ingest a large config, expand it out as protobuf for unknown field checking, even if most of the config is unused by Envoy or not fully instantiated. I wonder if we should be using protobuf with temporary arenas in these situations (is this even possible?).

Reply:
@htuch 
Do you mean to use temporary memory allocation for these scenarios instead of TCmalloc?

There are some scenarios where memory allocation is less frequent that I think can be handled in a temporary arenas way, such as config_dump. But in the xDS scenario, it's clear that it's not a low-frequency memory allocation. In this scenario, we still need to use efficient memory allocators.

Reply:
OK this makes sense to me, though it seems unfortunate that tcmalloc cannot be directly configured to do this. IIRC there is some env var in tcmalloc that theoretically controls how often tcmalloc gives memory back. Is it worth it to briefly look at the code to see if there is a way we could directly force this behavior vs. having to have this workaround? cc @jmarantz

Reply:
@mattklein123 In the actual production environment, I set the environment variable (TCMALLOC_RELEASE_RATE=10), but the effect was not obvious.

Reply:
OK yeah, I thought there was some env var but I do recall it not working correctly when I've tried it in the past, but I would not treat this as definitive.

Reply:
@jmarantz I am looking forward to your advice

Reply:
Hi -- sorry I missed this question. I don't have any more insight into tcmalloc's controls. It'd be nice to have some benchmarks on Envoy system performance so we could experiment with jemalloc and/or tcmalloc settings.

RE the bursts of memory required for some admin points; I do think `/stats` is pretty bad for that. It might be worth trying to stream the response out or deliver it in pages somehow, rather than buffering all of it. I haven't looked at `/clusters`'. But contribution toward that would be welcome.

I don't fully understand the other scenario where a bug in your control-plane caused a memory increase in Envoy. You are saying that such control-plane issues happen and Envoy should recover quicker when they do?

Reply:
@jmarantz Istio initially launched a large cluster, but after a while, the cluster went offline, and the Envoy's memory did not decrease

Reply:
Could we just add a utility `tryShrinkHeap()` and call this after known problematic admin endpoint and config ingestion sites? I think we could cover most of the problematic sites with this. Given this happens on the main thread, it shouldn't be too disruptive. Seems a more direct way to tackle the issue.

Reply:
@htuch cc @mattklein123 @jmarantz 
OK, I agree with this approach. I will close this PR and start a new issue to follow up.